opnode/p2p/config.go

```diff
@@ -0,0 +1,394 @@
+	for i, recordB64 := range records {
+		nodeRecord, err := enode.Parse(enode.ValidSchemes, strings.TrimSpace(recordB64))
+		if err != nil {
+			return fmt.Errorf("bootnode record %d (of %d) is invalid: %q err: %v", i, len(records), recordB64, err)
+		}
+		conf.Bootnodes = append(conf.Bootnodes, nodeRecord)
+	}
+
+	return nil
+}
+
+func yamuxC() (lconf.MsMuxC, error) {
+	mtpt, err := lconf.MuxerConstructor(yamux.DefaultTransport)
+	if err != nil {
+		return lconf.MsMuxC{}, err
+	}
+	return lconf.MsMuxC{MuxC: mtpt, ID: "/yamux/1.0.0"}, nil
+}
+
+func mplexC() (lconf.MsMuxC, error) {
+	mtpt, err := lconf.MuxerConstructor(mplex.DefaultTransport)
+	if err != nil {
+		return lconf.MsMuxC{}, err
+	}
+	return lconf.MsMuxC{MuxC: mtpt, ID: "/mplex/6.7.0"}, nil
+}
+
+func noiseC() (lconf.MsSecC, error) {
+	stpt, err := lconf.SecurityConstructor(noise.New)
+	if err != nil {
+		return lconf.MsSecC{}, err
+	}
+	return lconf.MsSecC{SecC: stpt, ID: noise.ID}, nil
+}
+
+func tlsC() (lconf.MsSecC, error) {
+	stpt, err := lconf.SecurityConstructor(tls.New)
+	if err != nil {
+		return lconf.MsSecC{}, err
+	}
+	return lconf.MsSecC{SecC: stpt, ID: tls.ID}, nil
+}
+
+func (conf *Config) loadLibp2pOpts(ctx *cli.Context) error {
+
+	addrs := strings.Split(ctx.GlobalString(flags.StaticPeers.Name), ",")
+	for i, addr := range addrs {
+		a, err := multiaddr.NewMultiaddr(addr)
+		if err != nil {
+			return fmt.Errorf("failed to parse multi addr of static peer %d (out of %d): %q err: %v", i, len(addrs), addr, err)
+		}
+		conf.StaticPeers = append(conf.StaticPeers, a)
+	}
+
+	for _, v := range strings.Split(ctx.GlobalString(flags.HostMux.Name), ",") {
+		v = strings.ToLower(strings.TrimSpace(v))
+		var mc lconf.MsMuxC
+		var err error
+		switch v {
+		case "yamux":
+			mc, err = yamuxC()
+		case "mplex":
+			mc, err = mplexC()
+		default:
+			return fmt.Errorf("could not recognize mux %s", v)
+		}
+		if err != nil {
+			return fmt.Errorf("failed to make %s constructor: %v", v, err)
+		}
+		conf.HostMux = append(conf.HostMux, mc)
+	}
+
+	secArr := strings.Split(ctx.GlobalString(flags.HostMux.Name), ",")
+	for _, v := range secArr {
+		v = strings.ToLower(strings.TrimSpace(v))
+		var sc lconf.MsSecC
+		var err error
+		switch v {
+		case "none": // no security, for debugging etc.
+			if len(conf.HostSecurity) > 0 || len(secArr) > 1 {
+				return errors.New("cannot mix secure transport protocols with no-security")
+			}
+			conf.NoTransportSecurity = true
+		case "noise":
+			sc, err = noiseC()
+		case "tls":
+			sc, err = tlsC()
+		default:
+			return fmt.Errorf("could not recognize security %s", v)
+		}
+		if err != nil {
+			return fmt.Errorf("failed to make %s constructor: %v", v, err)
+		}
+		conf.HostSecurity = append(conf.HostSecurity, sc)
+	}
+
+	conf.PeersLo = ctx.GlobalUint(flags.PeersLo.Name)
+	conf.PeersHi = ctx.GlobalUint(flags.PeersHi.Name)
+	conf.PeersGrace = ctx.GlobalDuration(flags.PeersGrace.Name)
+
+	// TODO parse NAT option
```

Comment:
Important for now or not?

Reply:
This is an old TODO, will clean it up. Idea was to offer more advanced NAT configuration, but we already have a flag for setting the advertised external IP explicitly, and I don't think anyone cares to select PMP, UPNP or both, as long as it can be disabled/enabled.